habitat/core/dataset.py

```diff
@@ -324,40 +342,100 @@ def __next__(self):
 
         :return: next episode.
         """
+        self._switch_scene_if()
 
         next_episode = next(self._iterator, None)
         if next_episode is None:
             if not self.cycle:
                 raise StopIteration
+
             self._iterator = iter(self.episodes)
+
             if self.shuffle:
-                self._shuffle_iterator()
+                self._shuffle()
+
             next_episode = next(self._iterator)
 
-        if self._prev_scene_id == next_episode.scene_id:
-            self._rep_count += 1
         if (
-            self.max_scene_repetition > 0
-            and self._rep_count >= self.max_scene_repetition - 1
+            self._prev_scene_id != next_episode.scene_id
+            and self._prev_scene_id is not None
         ):
-            self._shuffle_iterator()
             self._rep_count = 0
+            self._step_count = 0
 
         self._prev_scene_id = next_episode.scene_id
         return next_episode
 
-    def _shuffle_iterator(self) -> None:
+    def _switch_scene(self) -> None:
+        r"""Internal method to switch the scene
+        """
+        grouped_episodes = [
+            list(g)
+            for k, g in groupby(self._iterator, key=lambda x: x.scene_id)
+        ]
+
+        if len(grouped_episodes) > 1:
+            # Ensure we swap by moving the current group to the end
+            if self.shuffle:
+                random.shuffle(grouped_episodes[1:])
+
+            grouped_episodes = grouped_episodes[1:] + grouped_episodes[0:1]
+
+        self._iterator = iter(sum(grouped_episodes, []))
```

Comment:
Will the iterator reach the end anytime? Looks like it will be reset each time.

Reply:
`grouped_episodes` is computed by running out `self._iterator`, thus the length of `self._iterator` before and after this function will be exactly the same and it will eventually hit the end.